src/psyclone/psyGen.py

```diff
@@ -4486,6 +4563,78 @@ def update(self):
         self._add_region(start_text="DATA", end_text="END DATA",
                          data_movement="analyse")
 
+    def begin_string(self):
+        '''Returns the beginning statement of this directive, i.e.
+        "acc data". The visitor is responsible for adding the
+        correct directive beginning (e.g. "!$").
+
+        :returns: the opening statement of this directive.
+        :rtype: str
+
+        :raises NotImplementedError: if the region contains one or more \
+                                     references to structures (TODO #1028).
+
+        '''
+        result = "acc data"
+
+        struct_accesses = self.walk(StructureReference)
+        if struct_accesses:
+            # TODO #1028. Dependence analysis does not yet work for structure
+            # references.
+            # Have to import here to avoid circular dependency
+            # pylint: ignore=import-outside-toplevel
+            from psyclone.psyir.backend.fortran import FortranWriter
+            fwriter = FortranWriter()
+            ref_list = [fwriter(ref) for ref in struct_accesses]
+            raise NotImplementedError(
+                "Structure (derived-type) references are not yet supported "
+                "within OpenACC data regions but found: {0}".format(ref_list))
+
+        # Identify the inputs and outputs to the region (variables that
+        # are read and written).
+        from psyclone.core.access_info import VariablesAccessInfo
+        var_accesses = VariablesAccessInfo(self)
+        readers = set()
+        writers = set()
+        for var in var_accesses.all_vars:
+            accesses = var_accesses[var]
+            if not accesses.all_accesses[0].indices:
```

Comment:
I don't know enough about the overall code structure, but in general I try to deprecate the ``indices`` method, and instead use the symbol-table, e.g. (from dependency_tools):
```
            symbol_table = loop.scope.symbol_table
            if var_name not in symbol_table:
                # TODO #845: Once we have symbol tables, any variable should
                # be in a symbol table, so we have to raise an exception here.
                # We need to fall-back to the old-style test, since we do not
                # have information in a symbol table. So check if the access
                # information stored an index:
                is_array = var_info[0].indices is not None
            else:
                # Find the symbol for this variable
                symbol = symbol_table.lookup(var_name)
                is_array = symbol.is_array
```
This should much better handle the case if there is an array expression without ``:`` (eg. not ``a(:) = b(:) * 2``, but just ``a=b*2`` - the latter case would return no indices since at that time iirc we had no array representation that I could use).

Reply:
Oh yes, good call. I've done that now.